### ISSUE TITLE
SEP 038 -- SBOL3 Identity and URIs

### ISSUE THREAD

**Issue by jakebeal** [SynBioDex/SEPs#85]
This SEP proposes to modify the relationship between Identified objects and URIs to simplify name management and to disentangle object identity from knowledge graph versioning.

Full proposal in: https://github.com/SynBioDex/SEPs/blob/master/sep_038.md

**Comment by cjmyers** [comment]
I have a few questions about this proposal:
1) Is a namespace usually just a domain, like https://synbiohub.org?
2) What is the purpose of the Namespace object?  Is it simply to indicate what part of the URI is the namespace?  
3) When would you use root vs. collection_structure?
4) Having multiple versions of an object within the same collection_structure is not possible.  Isn't this kind of limiting?
5) Why does the child counter need to increment even for different types of objects since they already differ by the identifier part?  Also, why do we need to recommend this?  I think we can suggest it, but a recommendation is a best practice which I think is too strong.  Not sure what we gain in this case by child URIs having this compliant form.
6) You restrict out delimiters from displayIds, but they are already much more limited than that.  Are you suggesting to remove the limitations to alphanumeric and underscores?

**Comment by hasanbaig** [comment]
Hi guys,
There is not much activity on this thread. Lets begin the discussion to bring this SEP to conclusion before HARMONY 2020.

**Comment by jakebeal** [comment]
@cjmyers

- 1,2,3: I've updated the SEP to clarify the use of namespaces, addressing 1, 2, and 3.
- 4: I've removed versions entirely from recommendation and explained how there's freedom to operate across different potential versioning conventions.
- 5: I've seen people get terribly mucked up in trying to figure out how to name their child objects, coming up with lots of problematic schemas that don't guarantee non-collision.  If we specify a fixed algorithm, then it can be put into the libraries.  If we wanted to make it so that it increments only for the same type of object, that would be OK by me --- I just want a single recommended algorithm.
- 6. I wasn't intending to weaken the limitations.

**Comment by cjmyers** [comment]
In your example, is "/2017_distribution/" namespace or collection?

Couldn't SBOL2 objects, if needed, have the version become part of the collection structure?

BTW: versions were never required for compliant URIs.  They were only required to be included in the URI in a particular way, if a version is provided.  I'm okay with how versions are handled now in this SEP.

As for 5, I still don't see why numbering cannot be by type.  Not sure the risk there.

As for 6, you might make this clear in the SEP.

Discussion section is not complete sentence.

**Comment by jakebeal** [comment]
Whoops --- forgot to put that component in.  I've stuck it in the namespace.

If an application wants to preserve version, it can, but there's no guarantee of round-tripping between SBOL 2 and SBOL 3 for different tools for anything other than version 1.  I think this is OK because versions other than 1 are quite rare, and if we force round-tripping, then we force version structure in a URI in SBOL 3, which this SEP aims to avoid.

For 5, I did say I'm also OK with having numbering be by type.  Either by type or by all children is fine by me, as long as we pick one.

6: done.

**Comment by cjmyers** [comment]
So, I assume then you need the Namespace object to make it clear that it is part of the namespace and not the collection, right?

As for version, I'm just thinking about SBH migration issues.

My preference is to number by type.  SBOL3 serialization is not necessarily going to be nested, so it is going to be easier to have an idea looking at files about the numbers of certain types if numbered in this way.  

As for 6, maybe rephrase: "Since displayId's are restricted to alphanumeric and underscores, they will never include delimiters such as '#', ':', and '/'."

**Comment by cjmyers** [comment]
Otherwise, I think this is fine now.

**Comment by jakebeal** [comment]
Yes, the namespace is intended to differentiate where the collection portion of the URI starts. This will allow prefix rewriting to be very simple, because it will change to be just namespace rewriting.

For SBH, you can migrate versions however you want, since the spec doesn't constrain it.

I've made changes to address the other two, per request.

**Comment by cjmyers** [comment]
Ok, sounds good to me.

**Comment by hasanbaig** [comment]
Can we move on this SEP to voting?

**Comment by jakebeal** [comment]
Yes (thought we also anticipate some clarifying notes regarding UUID cases from @udp and @PrashantVaidyanathan )

**Comment by cjmyers** [comment]
Do UUIDs meet the required URI structure?  One of the big points of this SEP is to restrict URIs to be easier to manage in software, since general URIs are too variable.  I would object to UUIDs if these reintroduce the problems we currently have with open URIs.

**Comment by jakebeal** [comment]
My hope is that UUIDs will turn out to be a degenerate case of namespace, but I'm waiting for the examples to find out.

**Comment by cjmyers** [comment]
From here:

https://en.wikipedia.org/wiki/Universally_unique_identifier

The appear to look like this:

urn:uuid:123e4567-e89b-12d3-a456-426655440000

Not excited about this option.

**Comment by jakebeal** [comment]
Are they even compatible with SBOL 2 format?

@PrashantVaidyanathan, can you please give more information on how you are using UUIDs at MSR?

**Comment by cjmyers** [comment]
URIs are fairly permissive, so I'm pretty sure these would parse fine.  They would though likely cause some fits to some tooling that has assumptions about URI form.  My original understanding of this SEP was that it was going to place some restrictions on URIs while also cleaning up some problems with compliant URI definitions.  If we allow these, then I think we are perhaps back to allowing permissive URIs again.  

BTW: for namespace, we do not put any restrictions currently in this SEP on the protocol.  Namely, must all URIs start with "http" or "https".  Do we want such a requirement?  Otherwise, even no protocol is allowed.  Which would me "uri" would be a valid URI.

**Comment by udp** [comment]
It’s not really a compliant URIs problem.  UUIDs are the standard for identifiers in databases so IMO we should not disallow identifying an SBOL object with a UUID.

Using UUIDs is also extremely simple from a library developer perspective, because the logic for generating the URI of a child object to a top level labelled with a UUID is simply to create another UUID.  (We could have a rule that any child object of a top level identified by a UUID must also be identified with a UUID so that there is no ambiguity, which would also separate this issue entirely from the issue of compliant URIs).

**Comment by cjmyers** [comment]
My vote would be that while people are free to use UUIDs or other URI forms internally during development, that published URIs to the community should be compliant.

**Comment by cjmyers** [comment]
I'm still concerned though about tool interoperability issues, in any case with UUIDs, but if tools that use them can mint compliant URIs on export, then it does not really matter what they use internally.

**Comment by jakebeal** [comment]
@udp Right now, because the SEP says things that assume URL structure, it would disallow URIs.

Would it be sufficient to change the statement on "Required identity URI structure" to "Required identity URL structure for all cases where a URI is a URL"?

**Comment by udp** [comment]
Sure, I don’t want to see parts with UUIDs on synbiohub!  I just want an SBOL file that uses UUIDs to be allowed by the spec.

**Comment by udp** [comment]
That was in reply to Chris - we commented at the same time.  Jake, I think that sounds sensible.

**Comment by cjmyers** [comment]
@jakebeal sounds good.  Does this also mean that namespace is going to be required to start with "http://" or "https://" when it is URI publicly exposed as a URL?

**Comment by jakebeal** [comment]
@cjmyers While I expect that will generally be the case, I don't think that's strictly required. 

There will definitely be times that we want to use an externally-pointed URL that's aimed at a different type of system, such as doi (for references) or various file systems for large datasets (e.g., ftp, file, afs, agave). Since these generally aren't SBOL objects, they generally won't need to be in a NameSpace.

Given that we're allowing such URLs, though, I don't see any particular reason to explicitly exclude the possibility of using one in a NameSpace if it fits the format.

**Comment by jakebeal** [comment]
I've made the change in the SEP, and added some explanation as well.

Do note that UUIDs cannot be "compliant", but that's no different than it's ever been, and I think appropriate.  If you want to use UUIDs, you accept that certain tooling functionality may not apply for those documents.

**Comment by cjmyers** [comment]
We might consider at least requiring a protocol.  However, I think the namespace scheme may solve a lot of the problems from before, where it was just impossible to figure out which part was the namespace.  So, I think this is fine for now then.

**Comment by PrashantVaidyanathan** [comment]
We typically use GUIDs - The format is specified in the wikipedia article Chris mentioned. So it would be represented as: urn:uuid:123e4567-e89b-12d3-a456-426655440000

That being said, we can implement a work-around such that the GUIDs form the display. With that, we can create SBOL complaint URIs. So interoperability will not be much of an issue for us.

**Comment by jakebeal** [comment]
Excellent: so then it seems like the compromise now in the SEP will work: you can use UUIDs where you want (since they aren't constrained), and put in URLs when you want other SBOL tooling to be able to work with your outputs.
